checks/raw/pinned_dependencies.go

```diff
@@ -61,9 +61,117 @@ func PinningDependencies(c *checker.CheckRequest) (checker.PinningDependenciesDa
+	for i := 0; i < len(nugetDependencies); i++ {
+		if !*nugetDependencies[i].Pinned {
+			unpinnedNugetDependencies = append(unpinnedNugetDependencies, nugetDependencies[i])
+		}
+	}
+	return unpinnedNugetDependencies
+}
+
+func processCsprojLockedMode(c *checker.CheckRequest, dependencies []*checker.Dependency) error {
+	csprojDeps, err := collectCsprojLockedModeData(c)
+	if err != nil {
+		return err
+	}
+	unlockedCsprojDeps, unlockedPath := countUnlocked(csprojDeps)
+
+	// none of the csproject files set RestoreLockedMode. Keep the same status of the nuget dependencies
+	if unlockedCsprojDeps == len(csprojDeps) {
+		return nil
+	}
+
+	// all csproj files set RestoreLockedMode, update the dependency pinning status of all nuget dependencies to pinned
+	if unlockedCsprojDeps == 0 {
+		PinAllNugetDependencies(dependencies)
+	} else {
+		// only some csproj files are locked, keep the same status of the nuget dependencies but create a remediation
+		for i := 0; i < len(dependencies); i++ {
+			(dependencies)[i].Remediation = &finding.Remediation{
+				Text: (dependencies)[i].Remediation.Text +
+					": some of your csproj files set the RestoreLockedMode property to true, " +
+					"while other do not set it: " + unlockedPath,
+			}
+		}
+	}
+	return nil
+}
+
+func PinAllNugetDependencies(dependencies []*checker.Dependency) {
+	for i := 0; i < len(dependencies); i++ {
+		dependencies[i].Pinned = asBoolPointer(true)
+		dependencies[i].Remediation = nil
+	}
+}
+
+func collectCsprojLockedModeData(c *checker.CheckRequest) ([]checker.DotnetCsprojLockedData, error) {
+	var csprojDeps []checker.DotnetCsprojLockedData
+	if err := fileparser.OnMatchingFileContentDo(c.RepoClient, fileparser.PathMatcher{
+		Pattern:       "*.csproj",
+		CaseSensitive: false,
+	}, analyseCsprojLockedMode, &csprojDeps, c.Dlogger); err != nil {
+		return nil, err
+	}
+
+	return csprojDeps, nil
+}
+
+func analyseCsprojLockedMode(path string, content []byte, args ...interface{}) (bool, error) {
+	pdata, ok := args[0].(*[]checker.DotnetCsprojLockedData)
+	if !ok {
+		// panic if it is not correct type
+		panic(fmt.Sprintf("expected type *[]checker.DotnetCsprojLockedData, got %v", reflect.TypeOf(args[0])))
+	}
+
+	err, pinned := fileparser.IsRestoreLockedModeEnabled(content)
+	if err != nil {
+		dl, ok := args[1].(checker.DetailLogger)
+		if !ok {
+			// panic if it is not correct type
+			panic(fmt.Sprintf("expected type checker.DetailLogger, got %v", reflect.TypeOf(args[1])))
+		}
+
+		dl.Warn(&checker.LogMessage{
+			Text: fmt.Sprintf("malformed csproj file: %v", err),
+		})
+		return true, nil
+	}
+
+	csproj := checker.DotnetCsprojLockedData{
+		Path:          &path,
+		LockedModeSet: asBoolPointer(pinned),
+	}
+
+	*pdata = append(*pdata, csproj)
+	return true, nil
+}
+
+func countUnlocked(csprojFiles []checker.DotnetCsprojLockedData) (int, string) {
+	count := 0
+	unlockedCsProjPath := ""
```

Comment:
instead of managing the string joining yourself, lets use a slice and [strings.Join](https://pkg.go.dev/strings#Join) which will handle the edge cases for you and handle efficient string concatenation.

```go
var unlockedPaths []string

for i := range csprojFiles {
	if !*csprojFiles[i].LockedModeSet {
		unlockedPaths = append(unlockedPaths, csprojFiles[i].Path)
	}
}
return len(unlockedPaths), strings.Join(unlockedPaths, ", ")
```

Reply:
great stuff, thanks!!